### ISSUE TITLE
Add a probe preloading check at start-up to test kernel features

### ISSUE THREAD

**Issue by kakkoyun** [parca-dev/parca-agent#896]
libbpf has this nifty function to test kernel capabilities at loading time; however, the function is not exported, so it can't be used with libbpfgo directly https://github.com/libbpf/libbpf/blob/1714037104da56308ddb539ae0a362a9936121ff/src/libbpf.c#L4396-L4429

But we can add a similar check on our own by building a tiny probe program that does not do anything and check the kernel capabilities at start-up.

**Comment by maxbrunet** [comment]
Those checks do not run when the agent starts and loads libbpf?

For example, the agent bumps `rlimit_memlock`, and libbpfgo was doing it as well, but it has been removed (aquasecurity/libbpfgo#225), because it is automatically done in v1. Would it possible to remove it here and let libbpf handle that?

https://github.com/parca-dev/parca-agent/blob/1ce2d9ac8796622d4059fc05887231c8e22ceab5/pkg/profiler/cpu/cpu.go#L167-L172

**Comment by kakkoyun** [comment]
> Those checks do not run when the agent starts and loads libbpf?

Yes, but it is not in start-up and it only fails the profiler. We want to stop agent to work if we can't run eBPF programs.

> For example, the agent bumps `rlimit_memlock`, and libbpfgo was doing it as well, but it has been removed ([aquasecurity/libbpfgo#225](https://github.com/aquasecurity/libbpfgo/pull/225)), because it is automatically done in v1. Would it possible to remove it here and let libbpf handle that?

For the rlimit_memlock, we can remove it after migrating to libbpf v1.0.0. I haven't checked the implementation, does it bump the limit to match the map capacity?

**Comment by maxbrunet** [comment]
> does it bump the limit to match the map capacity?

https://github.com/libbpf/libbpf/commit/216eaa760edf06f3521d5d49a419515607c1f38e

TL;DR: Kernel <5.11, infinity. Kernel >5.11, nothing/unnecessary

**Comment by v-thakkar** [comment]
JFYI, there is a separate issue for the rlimit thingy https://github.com/parca-dev/parca-agent/issues/388.

**Comment by kakkoyun** [comment]
To clarify things here; The goal was to make sure the Kernel has all the necessary bits. The helper function that I've referred to happened to have the rlimit access check as well. So the goal of the issue is not the rlimit!

And it's time to address #388 actually, we have migrated to libbpf v1.0.0 already.

**Comment by kakkoyun** [comment]
@javierhonduco and I discussed having a program that would be loaded at startup to cover needed functionality. e.g. https://github.com/javierhonduco/rbperf/blob/main/src/bpf/features.bpf.c

Since the helper function that has been proposed in the description is not exported.

**Comment by v-thakkar** [comment]
Ah, yes. Forgot that we did it already. :see_no_evil:  I'll take care of #388.

> To clarify things here; The goal was to make sure the Kernel has all the necessary bits. The helper function that I've referred to happened to have the rlimit access check as well. So the goal of the issue is not the rlimit!

Yes, I got that. That's why I referenced rlimit one to continue the discussion there. Just to keep things separate. :)

> And it's time to address #388 actually, we have migrated to libbpf v1.0.0 already.

Ah, yes. Forgot that we did it already. :see_no_evil:  I'll take care of #388.
